---
title: "Mimicry_Orthogroups"
author: "Emily Phelps"
date: "`r Sys.Date()`"
output: html_document
---

```{r}
source("scripts/general/packages.R")
source("scripts/general/colours.R")
```
Get orthogroups uusing proteins from transcriptomes that are filtered for those over 50 amino acids and for those that are more than 90% different (CDHIT)
```{r}
ortho_count <- read_tsv("data/orthgroups/Orthogroups.GeneCount.tsv")

ortho_count <- ortho_count %>%
  mutate(across(c(ARC_renamed, MET_renamed, NAR_renamed, REY_renamed, SIM_renamed, 
  TUK_renamed), as.numeric))

# Filter out rows with any 0 in the specified columns
ortho_count_filtered <- ortho_count %>%
  filter(if_all(c(ARC_renamed, MET_renamed, NAR_renamed, REY_renamed, SIM_renamed, 
  TUK_renamed), 
                ~ . != 0))

orthologs <- read.table("data/orthgroups/Orthogroups.txt", sep=":", header = FALSE)

ortho.filt <- orthologs %>% 
  as_tibble() %>% 
  rename(V1="Orthogroup", V2="genes") %>% 
  filter(Orthogroup %in% ortho_count_filtered$Orthogroup) %>%
  separate(genes, paste0("gene", seq(1, 1000)), sep=" ") %>% 
  select(!gene1) %>%
  pivot_longer(cols=paste0("gene", seq(2, 1000)), values_to = "value", names_to = "name") %>%
  select(!name) %>% filter(!is.na(value)) %>% 
  mutate(species=substr(value, 1, 3),
         gene= substr(value, 4, str_length(value))) %>%
  mutate(gene=gsub("_i.*", "", gene))


```
```{r}
#source("/Users/emilyphelps/Documents/T/Toxin_Project/Venom_Evolution_2024/ortholog_map_creation.R")
```
```{r}
relationships <- read.table("samples.txt", header=F, sep="\t") %>% mutate(species=substr(V1, 1, 3))
colnames(relationships) <- c("sample", "species") 

species <- c("MET", "SIM", "REY", "TUK", "NAR", "ARC")
mimicry <- c("mim1", "mim1", "mim2", "mim2", "mim3", "mim3")
lineage <- c("L9", "L1", "L1", "L9", "L1", "L9")

relationships <- cbind(species, mimicry, lineage) %>% as.tibble() %>% left_join(relationships, .) %>%
                  mutate(tissue=ifelse(str_detect(pattern="_S", sample), "S", "A")) %>% filter(tissue == "A")
```

```{r}
txi <- list()
tx2gene.df <- data.frame()

for (i in species){
  rel <- relationships %>% filter(species == paste0(i)) 
  sf.files <- file.path("data/trinity_salmon_qual", rel$sample, "quant.sf")
  names(sf.files) <- as.vector(rel$sample)
 
  txi <- tximport(sf.files, type = "salmon", txOut = TRUE)
  tx2gene <- read_tsv(paste0("data/transcriptomes/", i,"/Ortholog_Trinity.fasta.gene_trans_map"), 
                      col_names=FALSE) %>% 
    dplyr::rename(gene=X1, transcript=X2) %>%
    mutate(species=paste0(i)) %>% dplyr::select(transcript, gene, species)
  tx2gene.df <- rbind(tx2gene.df, tx2gene)
  txi.sum <- summarizeToGene(txi, tx2gene)
  var_name <- paste0(i, ".txi")
  assign(var_name, txi.sum)
}

txi.list <- list(MET.txi, SIM.txi, REY.txi, TUK.txi, NAR.txi, ARC.txi)
names(txi.list) <- paste0(species, ".txi")
```

```{r}
process_data <- function(obj) {
  # Transfer row names to a new column 'gene'
  df <- obj$counts %>% as.data.frame() %>% rownames_to_column(var="gene") %>% arrange(gene)
  return(df)
}

filtered_data_list <- lapply(txi.list, process_data)

#Normalise by replicates then normalize by species. So create a dds and estimate vst. Check the normalisation and filtering here...
combined_matrix <- purrr::reduce(filtered_data_list, function(x, y) left_join(x, y, by = "gene")) %>% 
                   column_to_rownames(var="gene") %>% 
                   mutate_all(~ if_else(is.na(.), 0, .)) %>% 
                   as.matrix()

ordered <- relationships %>% arrange(factor(sample, levels = colnames(combined_matrix)))


dds <- DESeqDataSetFromMatrix(round(combined_matrix), colData = ordered, design = ~ lineage)
keep <- rowSums(counts(dds) > 50) >= 8
dds <- dds[keep,]

norm.data <- vst(dds)

lineage <- plotPCA(object=norm.data, intgroup="lineage") + theme_classic()

lineage$data %>% dplyr::select(!lineage) %>% left_join(., relationships, by=c("name" = "sample")) %>%
ggplot() + 
  geom_point(aes(x=PC1, y=PC2, color=lineage, shape=species)) + 
  geom_convexhull(aes(x=PC1, y=PC2, fill=lineage), alpha=0.4) + 
  labs(x="PC1: 59% variance", y="PC2: 16% variance", colour="Genus", fill= "Genus", shape="Species") + 
  scale_fill_manual(values=c(scale2[8], scale2[4]),
                    breaks=c("L9", "L1"), 
                    labels=c("Hoplisoma", "Corydoras")) +
  scale_colour_manual(values=c(scale2[8], scale2[4]),
                    breaks=c("L9", "L1"), 
                    labels=c("Hoplisoma", "Corydoras")) +
  theme(legend.position = "top",
        legend.key=element_blank(),
        panel.border = element_rect(color = theme_cols[4], 
                                    fill = NA, 
                                    size = 0.8),
        panel.background = element_blank(),
        axis.title= element_text(size= 10, color=theme_cols[4]), 
        axis.line = element_line(linewidth = 0.5, 
                                 colour = theme_cols[3]),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank())

ggsave("MS_lineage_pca.pdf", width=6, height=5)
```

```{r}
express <- DESeq(dds)
res <- results(express, alpha= 0.05)
resOrdered <- res[order(res$padj),]
summary(resOrdered)
resOrdered %>% as.data.frame() %>% rownames_to_column(var="gene") %>% filter(padj < 0.05) %>% write_tsv("MS_Lin_significant_genes.tsv")
VolcanoPlot(data = as.data.frame(res))
```
Need to get the orthogroups from SIM to cross compare.
Need to change the module venom. Ensure these are the same after correcting the SIM venom assessment
```{r}
venom.candidates <- read_tsv("venom_ortholog.tsv")
mod.venom <- read_tsv("module_venom_candidates.tsv")
venom.candidates <- venom.candidates %>% filter(gene_id %in% mod.venom$gene_id)
mimicry_venom.cands <- resOrdered %>%
                            as.data.frame() %>% 
                            rownames_to_column(var="gene") %>% 
                            filter(gene %in% as.vector(venom.candidates$Orthogroup) & padj < 0.05) %>%
                            left_join(., venom.candidates, by=c("gene" = "Orthogroup"))
                            

#mimicry_venom.cands %>% write_tsv("mimicry_venom_cands.tsv")
```

```{r}
library(forcats)
heat.data <- assay(norm.data) %>%
  as.data.frame() %>% 
  rownames_to_column(var="gene") %>% 
  filter(gene %in% as.vector(mimicry_venom.cands$gene)) %>%
  left_join(., venom.candidates, by=c("gene" = "Orthogroup")) %>%
  pivot_longer(cols=MET_01_A1_1:ARC5_A2_1, names_to = "sample", values_to="norm.expression") %>%
  left_join(., relationships, by="sample") %>% 
  mutate(lineage=ifelse(lineage =="L1", "Corydoras", "Hoplisoma"))


heat.plot <-ggplot() + 
            geom_tile(heat.data, mapping=aes(x=sample, 
                          y=name, 
                          fill=norm.expression)) +
            scale_fill_gradient2(low=divergent[4], 
                                 mid=divergent[6], 
                                 high=divergent[9]) +
            labs(x= "Individual", y="Candidate Venom Genes", fill="Normalized\nExpression") +
            facet_wrap(~ lineage, scales = "free_x",) +
            theme_minimal() +
            theme(legend.position="top",
                  axis.text.x = element_text(angle = 75, vjust = 0.5, hjust=0.5),
                  axis.text.y=element_text(size=12),
                  axis.title=element_text(size=12, color=theme_cols[3]),
                  axis.ticks=element_blank(),
                  strip.text.x = element_text(
                  size = 12, color=theme_cols[4]),
                  strip.background = element_rect(
                  color=theme_cols[3], size=0.3, linetype="solid"))

ggsave("heatmap_mim_lineage.pdf", width=10, height=10)
```

